datalad/local/subdatasets.py

```diff
@@ -284,21 +281,33 @@ def _get_submodules(ds, paths, fulfilled, recursive, recursion_limit,
     repo = ds.repo
     if not GitRepo.is_valid_repo(dspath):
         return
+    # expand all test cases for the contains test in the loop below
+    # leads to ~20% speedup per loop iteration of a non-match
+    expanded_contains = [
+        # convert to strings, as sm_path below will be a string and
+        # we do not want repeated type-conversion
+        [c] + list(c.parents)
+        for c in (contains if contains is not None else [])
+    ]
     # put in giant for-loop to be able to yield results before completion
     for sm in _parse_git_submodules(ds.pathobj, repo, paths):
-        contains_hits = []
+        sm_path = sm['path']
+        contains_hits = None
         if contains:
-            contains_hits = [
-                c for c in contains if sm['path'] == c or sm['path'] in c.parents
-            ]
+            contains_hits = [c[0] for c in expanded_contains if sm_path in c]
```

Comment:
Okay, so as far as I can see, the two potential sources of speed ups are that it 1) drops the repeated 'path' lookup in the `sm` dict and 2) uses list's `__contains__` over `_PathParents.__contains__` (which is `Sequence.__contains__`).

I doubt no. 1 would be measurable in this context, but I think it's still good to do.  So I'd guess the performance gains are coming from no 2.

<details>
<summary>timeit runs, .parents versus list __contains__</summary>

```sh
echo "--- .parents"
python -m timeit \
       -n 3 \
       -s 'from pathlib import Path' \
       -s 'import string' \
       'path = Path(*string.ascii_lowercase)' \
       'for _ in range(int(4e4)): "X" in path.parents'

echo "--- parents, no attribute access"
python -m timeit \
       -n 3 \
       -s 'from pathlib import Path' \
       -s 'import string' \
       -s 'path = Path(*string.ascii_lowercase)' \
       'parents = path.parents' \
       'for _ in range(int(4e4)): "X" in parents'

echo "--- parents, as list"
python -m timeit \
       -n 3 \
       -s 'from pathlib import Path' \
       -s 'import string' \
       -s 'path = Path(*string.ascii_lowercase)' \
       'parents = list(path.parents)' \
       'for _ in range(int(4e4)): "X" in parents'

# --- .parents
# 3 loops, best of 5: 1.31 sec per loop
# --- parents, no attribute access
# 3 loops, best of 5: 1.33 sec per loop
# --- parents, as list
# 3 loops, best of 5: 144 msec per loop
```

</details>